containerd-shim-spin/src/engine.rs

```diff
@@ -0,0 +1,238 @@
+                        .await
+                        .with_context(|| format!("failed to create {:?}", cache_dir))?;
+                }
+
+                spin_loader::from_file(&manifest_path, files_mount_strategy).await
+            }
+            ResolvedAppSource::OciRegistry { locked_app } => Ok(locked_app),
+        }
+    }
+
+    fn update_locked_app(&self, _locked_app: &mut LockedApp) {
+        // TODO: Apply --env to component environments
+    }
+
+    async fn write_locked_app(&self, locked_app: &LockedApp, working_dir: &Path) -> Result<String> {
+        let locked_path = working_dir.join("spin.lock");
+        let locked_app_contents =
+            serde_json::to_vec_pretty(&locked_app).context("failed to serialize locked app")?;
+        tokio::fs::write(&locked_path, locked_app_contents)
+            .await
+            .with_context(|| format!("failed to write {:?}", locked_path))?;
+        let locked_url = Url::from_file_path(&locked_path)
+            .map_err(|_| anyhow!("cannot convert to file URL: {locked_path:?}"))?
+            .to_string();
+
+        Ok(locked_url)
+    }
+
+    async fn build_spin_trigger<T: spin_trigger::TriggerExecutor>(
+        &self,
+        working_dir: PathBuf,
+        app: LockedApp,
+    ) -> Result<T>
+    where
+        for<'de> <T as TriggerExecutor>::TriggerConfig: serde::de::Deserialize<'de>,
+    {
+        let locked_url = self.write_locked_app(&app, &working_dir).await?;
+
+        // Build trigger config
+        let loader = loader::TriggerLoader::new(working_dir.clone(), true);
+        let runtime_config = RuntimeConfig::new(PathBuf::from("/").into());
+        let mut builder = TriggerExecutorBuilder::new(loader);
+        builder
+            .hooks(StdioTriggerHook {})
+            .config_mut()
+            .wasmtime_config()
+            .cranelift_opt_level(spin_core::wasmtime::OptLevel::Speed);
+        let init_data = Default::default();
+        let executor = builder.build(locked_url, runtime_config, init_data).await?;
+        Ok(executor)
+    }
+}
+
+impl Engine for SpinEngine {
+    fn name() -> &'static str {
+        "spin"
+    }
+
+    fn run_wasi(&self, _ctx: &impl RuntimeContext, stdio: Stdio) -> Result<i32> {
+        info!("setting up wasi");
+        stdio.redirect()?;
+        let rt = Runtime::new().context("failed to create runtime")?;
+
+        rt.block_on(self.wasm_exec_async())?;
+        Ok(0)
+    }
+
+    fn can_handle(&self, _ctx: &impl RuntimeContext) -> Result<()> {
+        Ok(())
+    }
+}
+
+fn parse_addr(addr: &str) -> Result<SocketAddr> {
+    let addrs: SocketAddr = addr
+        .to_socket_addrs()?
+        .next()
+        .ok_or_else(|| anyhow!("could not parse address: {}", addr))?;
+    Ok(addrs)
+}
+
+pub enum ResolvedAppSource {
+    File {
+        manifest_path: PathBuf,
+        manifest: AppManifest,
+    },
+    OciRegistry {
+        locked_app: LockedApp,
+    },
+}
+
+impl ResolvedAppSource {
```

Comment:
Since these are useful concepts maybe we should open an issue in spin to get them exposed as part of the sdk?

Reply:
Sure happy to make a PR to spin

Reply:
Added a comment

Reply:
https://github.com/fermyon/spin/issues/2073